### DISCUSSION TITLE
Instance network metrics calculation with API

### DISCUSSION THREAD

**Discussion by meisammaleki** [apache/cloudstack#12727] [Q&A]
##### ISSUE TYPE

 * Other

##### COMPONENT NAME

~~~

~~~

##### CLOUDSTACK VERSION 4.17.2

~~~

~~~

##### SUMMARY
I need to calculate the amount of downloaded and uploaded traffic on each instance for accounting purpose
For example if Download value for an instance reach 20 GB in a week (Download a 1GB file for 20 times)

##### STEPS TO REPRODUCE

~~~

~~~

##### EXPECTED RESULTS
I have downloaded this file with "wget" which it is approximately 2 GB with an instance and I except that the API show me a same value but was not same acctually. here is the like for download file 

https://github.com/szalony9szymek/large/releases/download/free/large

this is returned values form api :
"networkread": "41.20 MiB",
"networkwrite": "0.91 MiB",

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

**Comment by DaanHoogland** [comment]
@meisammaleki can you try your scenario with a newer version?

**Comment by meisammaleki** [comment]
The only way for check newer version is to use ACS simulator which I'm not sure about its functionality, also I cannot upgrade the current environment to newer one

**Comment by DaanHoogland** [comment]
no @meisammaleki , in a simulator this functionality can not be verified. 

> The only way for check newer version is to use ACS simulator which I'm not sure about its functionality, also I cannot upgrade the current environment to newer one

I'll try and find the time to verify it in a newer version.

**Comment by weizhouapache** [comment]
What type of zone and network ?

ACS can collect traffic for
- isolated network and vpc. No traffic collection for individual vm
- vm on shared netwoek

**Comment by Pearl1594** [comment]
@meisammaleki Could you also please let us know which API are you referring to where you see the discrepancy? If you are referring to the `listVirtualMachineMetrics` API response, please pass the `accumulate` parameter and set it to true, to see the correct value. When you call the API without accumulate=true, the networkread value (e.g., "41.20 MiB") represents only the inbound traffic in the last 60 seconds - essentially the current download rate, not the cumulative total. This is why it appears much lower than expected. To see the correct total inbound traffic that your VM has received, you must add accumulate=true to your API call: cloudmonkey list virtualmachinemetrics id=<vm-uuid> accumulate=true. This will return the sum of all network traffic across the stats retention period.

**Comment by DaanHoogland** [comment]
this sounds like a functional discussion...
